File: src/c/weighted.c
#include <stdio.h>

void print_seq(int* p, int sz); 
void set_union(int* id, int* sz, int p, int q); 
int connected(int* id, int p , int q); 
int root(int* id, int n); 

int main()
{ 
	int id[10]; 
	int sz[10]; 

	for(int i = 0; i < 10; i++)
		id[i] = i; 
	for(int i = 0; i < 10; i++)
		sz[i] = 1; 

	set_union(id, sz, 4, 3); 
	set_union(id, sz, 3, 8); 
	set_union(id, sz, 6, 5); 
	set_union(id, sz, 9, 4); 

	printf("%d\n", connected(id, 8, 9)); 
	printf("%d\n", connected(id, 5, 4)); 

	set_union(id, sz, 5, 0); 
	set_union(id, sz, 7, 2); 
	set_union(id, sz, 6, 1); 
	set_union(id, sz, 7, 3); 

	printf("%d\n", connected(id, 5, 4)); 
	return 0; 
}

void print_seq(int* p, int sz)
{
	for(int i = 0; i < sz; i++)
		printf("%d ", *(p + i)); 	

	printf("\n"); 	
} 

void set_union(int* id, int* sz, int p, int q) 
{
	int rp = root(id, p); 
	int rq = root(id, q); 

	if(sz[rp] < sz[rq])
	{
		id[rq] = id[rp]; 
		sz[rp]++; 
	}
	else
	{
		id[rp] = id[rq]; 
		sz[rq]++; 
	}
	print_seq(sz, 10); 
}

int connected(int* id, int p, int q) 
{
	return root(id, p) == root(id, q); 
}

int root(int* id, int n)
{ 
	if(id[n] == n)
		return n; 
	else
		return root(id, id[n]); 
} 

File: src/java/Weighted.java
public class Weighted
{
	public int[] seq = new int[10]; 
	public int[] sz = new int[10]; 

	public static void main(String[] args)
	{
		Weighted q = new Weighted();
		q.init();

		q.setUnion(4, 3); 
		q.setUnion(3, 8); 
		q.setUnion(6, 5); 
		q.setUnion(9, 4); 

		System.out.println(q.connected(8, 9)); 
		System.out.println(q.connected(5, 4)); 

		q.setUnion(5, 0); 
		q.setUnion(7, 2); 
		q.setUnion(6, 1); 
		q.setUnion(7, 3); 

		System.out.println(q.connected(2, 3)); 
	}

	public void init()
	{
		for(int i = 0; i < 10; i++)
		{
			seq[i] = i; 
			sz[i] = 1; 
		}
	}

	public void setUnion(int p, int q)
	{
		if(sz[p] < sz[q])
		{
			seq[root(q)] = seq[p];
			sz[p]++;
		}
		else
		{
			seq[root(p)] = seq[q];
			sz[q]++;
		}
	}

	public boolean connected(int p, int q)
	{
		return root(p) == root(q);
	}

	public int root(int n)
	{
		return seq[n] == n ? n : root(seq[n]); 
	}
}


